File: frontend/src/types/assignment.ts
export interface AssignmentSubmissionBrief {
  id: number;
  project_title: string;
  team_identifier: string;
}

export interface AssignmentJudgeBrief {
  id: number;
  email: string;
  full_name: string | null;
}

export interface AssignmentOut {
  id: number;
  submission: AssignmentSubmissionBrief;
  judge: AssignmentJudgeBrief;
  source: "auto" | "manual";
  assigned_at: string;
  evaluation_status: string;
}

export interface GenerateAssignmentsRequest {
  judges_per_submission: number;
  reset_existing: boolean;
}

export interface GenerateAssignmentsResult {
  assignments_created: number;
  submissions_covered: number;
  judges_used: number;
  min_load: number;
  max_load: number;
  warnings: string[];
}

export interface ManualAssignmentCreate {
  submission_id: number;
  judge_id: number;
}

export interface ConflictExclusionCreate {
  submission_id: number;
  judge_id: number;
  reason?: string | null;
}

export interface ConflictExclusionOut {
  id: number;
  submission_id: number;
  judge_id: number;
  reason: string | null;
  created_at: string;
}

File: backend/app/schemas/assignment.py
from datetime import datetime

from pydantic import BaseModel, Field


class GenerateAssignmentsRequest(BaseModel):
    judges_per_submission: int = Field(gt=0, le=50)
    reset_existing: bool = Field(
        default=False, description="If true, delete all existing assignments before generating new ones."
    )


class GenerateAssignmentsResult(BaseModel):
    assignments_created: int
    submissions_covered: int
    judges_used: int
    min_load: int
    max_load: int
    warnings: list[str]


class ManualAssignmentCreate(BaseModel):
    submission_id: int
    judge_id: int


class AssignmentSubmissionBrief(BaseModel):
    id: int
    project_title: str
    team_identifier: str

    model_config = {"from_attributes": True}


class AssignmentJudgeBrief(BaseModel):
    id: int
    email: str
    full_name: str | None

    model_config = {"from_attributes": True}


class AssignmentOut(BaseModel):
    id: int
    submission: AssignmentSubmissionBrief
    judge: AssignmentJudgeBrief
    source: str
    assigned_at: datetime
    evaluation_status: str

    model_config = {"from_attributes": True}


class ConflictExclusionCreate(BaseModel):
    submission_id: int
    judge_id: int
    reason: str | None = Field(default=None, max_length=255)


class ConflictExclusionOut(BaseModel):
    id: int
    submission_id: int
    judge_id: int
    reason: str | None
    created_at: datetime

    model_config = {"from_attributes": True}
